### PR TITLE
fix: WCO window hover on window controls on Windows

### LINKED ISSUE #32360: [Bug]: Windows Control Overlay button hover effect not working
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a bug report that matches the one I want to file, without success.

### Electron Version

14.0.0

### What operating system are you using?

Windows

### Operating System Version

Windows 11

### What arch are you using?

x64

### Last Known Working Electron version

_No response_

### Expected Behavior

See https://github.com/electron/electron/issues/30764 for specific behavior

### Actual Behavior

See https://github.com/electron/electron/issues/30764 for specific behavior

### Testcase Gist URL

_No response_

### Additional Information

I am creating this issue based on https://github.com/electron/electron/issues/30764 and the information I have gathered while trying to triage it. Though this not a breaking bug, it does create a visual inconsistency for users, as a window control button would typically change color when hovered over. For this reason, I think it should still be addressed at some point. Consolidating all current information and future questions/comments here should help in finding a solution.

[Here](https://bugs.chromium.org/p/chromium/issues/detail?id=1218921) is a link to the bug that led to the original fix implemented in Chromium

And [here](https://bugs.chromium.org/p/chromium/issues/detail?id=1271293#c1) is the link to the follow-up bug report created noting issues with the implemented fix. This was then labeled as `WontFix` 

The chromium fix artificially creates the correct tooltips and allows for the button color to change on hover. However, this results in a response to the `WM_NCHITTEST` [message](https://docs.microsoft.com/en-us/windows/win32/inputdev/wm-nchittest) that does not correspond with the function of the button. For example, `HTMAXIMIZE` is not returned, thus Snap Layout on Windows 11 does not work.

If someone else wants to take a look at this, the relevant Electron files can be found in [this PR](https://github.com/electron/electron/pull/29600) adding WCO to Electron. For this bug in particular, I believe the related files are: `shell/browser/ui/views/win_caption_button.cc`, `shell/browser/ui/views/win_caption_button_container.cc`, and `shell/browser/ui/views/win_frame_view.cc` but changes to other files may provide a solution as well.

### PR DESCRIPTION
**Pull request by clavin**
#### Description of Change

This adds a patch to Chromium that marks [non-client](https://docs.microsoft.com/en-us/windows/win32/gdi/nonclient-area) mouse hover events as handled so they don't get redispatched to the default OS event handler (which is currently immediately "stealing" the mouse every time it moves).

Currently, on Windows with WCO-enabled windows, Win API mouse messages first find their way to `LegacyRenderWidgetHostHWND::OnMouseRange` which, for non-client mouse moved events, will both handle the event normally (through `HWNDMessageHandler::HandleMouseMessage` which eventually bubbles down to `Widget::OnMouseEvent`) as well as redispatch the original message back to the OS default window procedure (via the Win API `DefWndProc`) on the real window handle, depending on some conditions.

Calling `DefWndProc` after we process a non-client area mouse move message seems to cause issues where the system will immediately send a non-client mouse **exit** message to the real window (`HWNDMessageHandler` directly this time, which again flows down to `Widget::OnMouseEvent`) which makes it think the mouse is no longer hovering over the non-client area.

To avoid this, an event should be marked as handled meaning `LegacyRenderWidgetHostHWND` will no longer forward the message to `DefWndProc`. Note that this is how many other mouse events behave in that method, but not hovers currently.

The only open question I haven't figured out is why `DefWndProc` is taking a `WM_NCMOUSEMOVE` message from `LegacyRenderHostHWND` called on `HWNDMessageHandler`'s window handle and then sending a `WM_NCMOUSELEAVE` message (take note: _move_ changed to _leave_) when it invokes `HWNDMessageHandler`'s `WndProc`. My current thought is that we tell the system that the client area covers the full window (i.e. the window has no frame) but also indicate that the mouse is sometimes in non-client area via hit testing (which we have to do to get Snap Assist) and this sort of inconsistency might trick `DefWndProc` into the behavior we're seeing. Not really sure if it's worth confirming though so it might just stay a hypothesis.

In theory this shouldn't affect anything but WCO windows as usually the `LegacyRenderWidgetHostHWND` doesn't cover any non-client area, so I deduce that all other kinds of windows will never send the `LegacyRenderWidgetHostHWND` a non-client mouse message in practice.

Fixes #32360.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Fix effect when hovering over window controls on Windows in a WCO-enabled window.

### CODE REVIEW

**Comment by clavin** [comment]
I'm working on upstreaming this, either as-is or as a small refactor. Either way and whether it gets accepted there or not, I think it would be a good idea to land this as a patch for now rather than when it finds its way in through the roll!

**Review by zcbenz — APPROVED** [review]
Excellent fix!

**Comment by deepak1556** [comment]
Is it possible to signal the handling of this mouse event at `ElectronDesktopWindowTreeHostWin` layer via overriding [HandleMouseEvent](https://source.chromium.org/chromium/chromium/src/+/main:ui/views/win/hwnd_message_handler_delegate.h;l=194) which acts as a delegate to `HWNDMessageHandler` and should help avoid the patch.

**Comment by clavin** [comment]
Good catch! I know the `DesktopWindowTreeHostWin` does pass along this event, so we should be able to catch it in `ElectronDesktopWindowTreeHostWin` and set it handled there. I'll try it and then push the change!

Also I just realized when I cleaned up the patch I broke the condition 😅 (should be `!= 0` instead of `== 1`) whoops.

**Comment by clavin** [comment]
@deepak1556 That works! It took me a minute to figure out, but as long as we do it after `SendEventToSink` then it works perfectly. I'll cook up a better change now.

File: shell/browser/ui/win/electron_desktop_window_tree_host_win.cc [hunk 1]
```diff
@@ -99,4 +99,21 @@ bool ElectronDesktopWindowTreeHostWin::GetClientAreaInsets(
   return false;
 }
 
+bool ElectronDesktopWindowTreeHostWin::HandleMouseEvent(ui::MouseEvent* event) {
+  // Call the default implementation of this method to get the event to its
+  // proper handler.
+  bool handled = this->views::DesktopWindowTreeHostWin::HandleMouseEvent(event);
```
Comment (deepak1556) [inline shell/browser/ui/win/electron_desktop_window_tree_host_win.cc:105]:
```suggestion
  bool handled = views::DesktopWindowTreeHostWin::HandleMouseEvent(event);
```

**Review by deepak1556 — APPROVED** [review]
👍

**Pull request merged**